File: python/src/retry.py
# Opt-in 429 retry/backoff for the generated client. See ../RETRY_POLICY.md for the full policy and why
# it's safe to retry every HTTP method (including POST/PUT/DELETE) specifically for 429 - the API's rate
# limiter runs before any handler, so a 429 guarantees the request was never processed.
#
# Unlike the TypeScript/C# SDKs, no hand-written retry loop is needed here: urllib3's Retry already
# understands Retry-After (respect_retry_after_header) and exponential backoff (backoff_factor/backoff_max)
# natively. This is just a small, policy-scoped factory around it. Attach the result to the generated
# client's Configuration before constructing an ApiClient:
#
#   from ordereazi_commerce_api import Configuration
#   from retry import build_retry
#
#   config = Configuration(host="https://api.example.com")
#   config.retries = build_retry()

from urllib3.util.retry import Retry


def build_retry(max_retries: int = 3, backoff_factor: float = 0.5, max_backoff: float = 30.0) -> Retry:
    """Builds a Retry that retries ONLY on 429, honoring Retry-After when present and falling back to
    exponential backoff (backoff_factor * 2**(attempt-1), capped at max_backoff) otherwise.

    connect/read/redirect/other are explicitly zeroed so a connection-level failure is never retried by
    this policy - only 429 is, matching the TypeScript/C# SDKs exactly. raise_on_status=False so that
    exhausting max_retries surfaces the final 429 as a normal ApiException(status=429, ...) instead of
    urllib3 raising its own MaxRetryError.
    """
    return Retry(
        total=max_retries,
        status=max_retries,
        connect=0,
        read=0,
        redirect=0,
        other=0,
        status_forcelist=[429],
        allowed_methods=None,
        backoff_factor=backoff_factor,
        backoff_max=max_backoff,
        respect_retry_after_header=True,
        raise_on_status=False,
    )

File: typescript/src/retry.ts
/**
 * Opt-in 429 retry/backoff for the generated client. See ../../RETRY_POLICY.md for the full policy and why
 * it's safe to retry every HTTP method (including POST/PUT/DELETE) specifically for 429 - the API's rate
 * limiter runs before any handler, so a 429 guarantees the request was never processed.
 *
 * This wraps a plain axios instance with a response interceptor - it does NOT touch anything under
 * src/generated/, so it survives regeneration untouched. Pass the returned instance as the third constructor
 * argument to any generated `XxxApi` class (e.g. `new CartApi(config, basePath, createRetryingAxios())`).
 */
import axios, { type AxiosInstance, type AxiosError } from 'axios';

export interface RetryOptions {
	/** Maximum number of retries after the initial attempt. Default 3 (4 total attempts). */
	maxRetries?: number;
	/** Base delay in milliseconds used for the exponential-backoff fallback when no Retry-After header is present. Default 500. */
	baseDelayMs?: number;
	/** Upper bound in milliseconds for any single wait, regardless of what Retry-After or backoff computed. Default 30000. */
	maxDelayMs?: number;
}

const DEFAULTS: Required<RetryOptions> = {
	maxRetries: 3,
	baseDelayMs: 500,
	maxDelayMs: 30_000,
};

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}

function delayForAttempt(error: AxiosError, attempt: number, options: Required<RetryOptions>): number {
	const retryAfterHeader = error.response?.headers?.['retry-after'];
	if (retryAfterHeader !== undefined) {
		const seconds = Number(retryAfterHeader);
		if (!Number.isNaN(seconds) && seconds >= 0) {
			return Math.min(seconds * 1000, options.maxDelayMs);
		}
	}

	const backoff = options.baseDelayMs * 2 ** attempt;
	const jitter = Math.random() * options.baseDelayMs;
	return Math.min(backoff + jitter, options.maxDelayMs);
}

/**
 * Creates an axios instance that automatically retries on HTTP 429, honoring the `Retry-After` header when
 * present and falling back to exponential backoff with jitter otherwise. Every other status code (including
 * 5xx) passes through unchanged - see RETRY_POLICY.md for why 5xx isn't covered yet.
 */
export function createRetryingAxios(options: RetryOptions = {}): AxiosInstance {
	const resolved: Required<RetryOptions> = { ...DEFAULTS, ...options };
	const instance = axios.create();

	instance.interceptors.response.use(undefined, async (error: AxiosError) => {
		const config = error.config as (AxiosError['config'] & { __retryCount?: number }) | undefined;

		if (!config || error.response?.status !== 429) {
			return Promise.reject(error);
		}

		config.__retryCount = config.__retryCount ?? 0;
		if (config.__retryCount >= resolved.maxRetries) {
			return Promise.reject(error);
		}

		config.__retryCount += 1;
		await sleep(delayForAttempt(error, config.__retryCount, resolved));
		return instance.request(config);
	});

	return instance;
}
